pkg/sypgp/sypgp.go

```diff
@@ -741,136 +795,166 @@ func ExportPubKey(kpath string, armor bool) error {
 	}
-
 	return nil
 }
 
 // ImportPubKey Will import a public key from a file (kpath).
-func ImportPubKey(kpath string) error {
+func ImportPubKey(entity *openpgp.Entity) error {
+
 	// Load the local public keys as entitylist
 	publicEntityList, err := LoadPubKeyring()
 	if err != nil {
 		return err
 	}
-	// Load the public key as an entitylist
-	pathEntityList, err := LoadKeyringFromFile(kpath)
-	if err != nil {
-		return err
-	}
+
 	// Get local keystore (where the key will be stored)
 	publicFilePath, err := os.OpenFile(PublicPath(), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
 	if err != nil {
 		return err
 	}
 	defer publicFilePath.Close()
 
-	// Go through the keyring checking for the given fingerprint
-	for _, pathEntity := range pathEntityList {
-		isInStore := false
-		for _, publicEntity := range publicEntityList {
-			if pathEntity.PrimaryKey.KeyId == publicEntity.PrimaryKey.KeyId {
-				isInStore = true
-				// Verify that this key has already been added
-				break
-			}
+	isInStore := false
+	for _, publicEntity := range publicEntityList {
+		if entity.PrimaryKey.KeyId == publicEntity.PrimaryKey.KeyId {
+			isInStore = true
+			// Verify that this key has already been added
+			break
 		}
-		if !isInStore {
-			if err = pathEntity.Serialize(publicFilePath); err != nil {
-				return err
-			}
-			fmt.Printf("Key with fingerprint %X succesfully added to the keyring\n", pathEntity.PrimaryKey.Fingerprint)
-		} else {
-			fmt.Printf("The key you want to add with fingerprint %X already belongs to the keyring\n", pathEntity.PrimaryKey.Fingerprint)
+	}
+	if !isInStore {
+		if err = entity.Serialize(publicFilePath); err != nil {
+			return err
 		}
+		fmt.Printf("Key with fingerprint %X succesfully added to the keyring\n", entity.PrimaryKey.Fingerprint)
+	} else {
+		fmt.Printf("The key you want to add with fingerprint %X already belongs to the keyring\n", entity.PrimaryKey.Fingerprint)
 	}
+
 	return nil
 }
 
+func getTypesFromEntity(path string) []string {
+	var types []string
+
+	f, err := os.Open(path)
+	if err != nil {
+		return types
+	}
+	defer f.Close()
+
+	el, err := openpgp.ReadKeyRing(f)
+	if err != nil {
+		// is armored, so need to identify each of the block types and store them
+		re := ReformatGPGExportedFile(f)
+		block, err := armor.Decode(re)
+		if err != nil {
+			return types
+		}
+		types = append(types, block.Type)
+	}
+	// is not armored so obtain the types checking the privatekey field from entity
+	for _, pathEntity := range el {
+		if pathEntity.PrivateKey != nil {
+			types = append(types, PrivateKeyType)
+		} else {
+			types = append(types, PublicKeyType)
+		}
+	}
+
+	return types
+}
+
 // ImportKey Will import a key from a file, and decied if its
 // a public, or private key.
 func ImportKey(kpath string) error {
 
 	// Load the private key as an entitylist
 	pathEntityList, err := LoadKeyringFromFile(kpath)
+	pathEntityTypes := getTypesFromEntity(kpath)
```

Comment:
I would move this line right under the `if err != nil {...}`, so it will look like:


```go
	pathEntityList, err := LoadKeyringFromFile(kpath)
	if err != nil {
￼		return fmt.Errorf("unable to get entity from: %s: %v", kpath, err)
￼	}
￼
	pathEntityTypes := getTypesFromEntity(kpath)
￼

```